lib/utils/packaging/unarchive_unix.go

```diff
@@ -0,0 +1,166 @@
+//go:build !windows
+
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package packaging
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	"errors"
+	"io"
+	"log/slog"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"slices"
+	"strings"
+
+	"github.com/google/renameio/v2"
+	"github.com/gravitational/trace"
+)
+
+// Replace un-archives package from tools directory and replaces defined apps by symlinks.
+func Replace(toolsDir string, archivePath string, hash string, apps []string) error {
+	switch runtime.GOOS {
+	case "darwin":
+		return replacePkg(toolsDir, archivePath, hash, apps)
+	default:
+		return replaceTarGz(toolsDir, archivePath, apps)
+	}
+}
+
+func replaceTarGz(toolsDir string, archivePath string, apps []string) error {
+	tempDir := renameio.TempDir(toolsDir)
+	f, err := os.Open(archivePath)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	gzipReader, err := gzip.NewReader(f)
+	if err != nil {
+		return trace.NewAggregate(err, f.Close())
+	}
+
+	tarReader := tar.NewReader(gzipReader)
+	for {
+		header, err := tarReader.Next()
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		// Skip over any files in the archive that are not in apps.
+		if !slices.ContainsFunc(apps, func(s string) bool {
+			return strings.HasSuffix(header.Name, s)
+		}) {
+			if _, err := io.Copy(io.Discard, tarReader); err != nil {
+				slog.DebugContext(context.Background(), "failed to discard", "name", header.Name, "error", err)
+			}
+			continue
+		}
+		// Verify that we have enough space for uncompressed file.
+		if err := checkFreeSpace(tempDir, uint64(header.Size)); err != nil {
+			return trace.NewAggregate(err, gzipReader.Close(), f.Close())
```

Comment:
No need to capture every close error for readers -- these are generally inconsequential